File: sdk/python/symbolchain/symbol/KeyPairV2.py
"""V2 key pair and verifier supporting Ed25519 and post-quantum schemes."""

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ed25519

from .CryptoScheme import CryptoScheme, PRIVATE_KEY_SIZES
from .CryptoTypesV2 import PublicKeyV2, SignatureV2, PrivateKeyV2

# ML-DSA-65 key layout constants
_ML_DSA_65_OQS_SECRET_KEY_SIZE = 4032
_ML_DSA_65_PUBLIC_KEY_SIZE = 1952

# PQC backend registry
_pqc_backends = {}


def register_pqc_backend(scheme_id, backend):
	"""Registers a PQC signing backend for a given crypto scheme.

	Expected backend interface:
		generate_key_pair() -> (public_key: bytes, secret_key: bytes)
		sign(message: bytes, secret_key: bytes) -> bytes
		verify(message: bytes, signature: bytes, public_key: bytes) -> bool
	"""
	_pqc_backends[scheme_id] = backend


def _get_pqc_backend(scheme_id):
	if scheme_id not in _pqc_backends:
		raise RuntimeError(f'no PQC backend registered for scheme {scheme_id}; call register_pqc_backend() first')
	return _pqc_backends[scheme_id]


class KeyPairV2:
	"""Represents a V2 key pair supporting Ed25519 and post-quantum schemes."""

	def __init__(self, private_key):
		"""Creates a V2 key pair from a V2 private key (size determines scheme)."""
		if not isinstance(private_key, PrivateKeyV2):
			raise TypeError('private_key must be a PrivateKeyV2 instance')

		self._scheme = private_key.crypto_scheme
		self._private_key = private_key

		if self._scheme == CryptoScheme.ED25519:
			self._sk = ed25519.Ed25519PrivateKey.from_private_bytes(private_key.bytes)

	@property
	def crypto_scheme(self):
		"""Gets the crypto scheme identifier."""
		return self._scheme

	@property
	def public_key(self):
		"""Gets the public key."""
		if self._scheme == CryptoScheme.ED25519:
			raw = self._sk.public_key().public_bytes(serialization.Encoding.Raw, serialization.PublicFormat.Raw)
			return PublicKeyV2(raw)

		# ML-DSA-65: public key stored at offset OQS_SECRET_KEY_SIZE
		pk_offset = _ML_DSA_65_OQS_SECRET_KEY_SIZE
		return PublicKeyV2(self._private_key.bytes[pk_offset:pk_offset + _ML_DSA_65_PUBLIC_KEY_SIZE])

	@property
	def private_key(self):
		"""Gets the private key."""
		return PrivateKeyV2(self._private_key.bytes)

	def sign(self, message):
		"""Signs a message."""
		if self._scheme == CryptoScheme.ED25519:
			return SignatureV2(self._sk.sign(message))

		backend = _get_pqc_backend(self._scheme)
		oqs_sk = self._private_key.bytes[:_ML_DSA_65_OQS_SECRET_KEY_SIZE]
		return SignatureV2(backend.sign(message, oqs_sk))

	@staticmethod
	def generate(scheme_id=CryptoScheme.ED25519):
		"""Generates a new V2 key pair for the given crypto scheme."""
		if scheme_id == CryptoScheme.ED25519:
			sk = ed25519.Ed25519PrivateKey.generate()
			raw_sk = sk.private_bytes(
				encoding=serialization.Encoding.Raw,
				format=serialization.PrivateFormat.Raw,
				encryption_algorithm=serialization.NoEncryption()
			)
			return KeyPairV2(PrivateKeyV2(raw_sk))

		backend = _get_pqc_backend(scheme_id)
		pk, sk = backend.generate_key_pair()

		# Store in our format: sk || pk
		combined = sk + pk
		if len(combined) != PRIVATE_KEY_SIZES[scheme_id]:
			raise ValueError(f'generated key size mismatch: {len(combined)} != {PRIVATE_KEY_SIZES[scheme_id]}')
		return KeyPairV2(PrivateKeyV2(combined))


class VerifierV2:
	"""Verifies V2 signatures from any supported crypto scheme."""

	def __init__(self, public_key):
		"""Creates a V2 verifier from a V2 public key."""
		if not isinstance(public_key, PublicKeyV2):
			raise TypeError('public_key must be a PublicKeyV2 instance')

		if bytes(len(public_key.bytes)) == public_key.bytes:
			raise ValueError('public key cannot be zero')

		self.public_key = public_key
		self._scheme = public_key.crypto_scheme

		if self._scheme == CryptoScheme.ED25519:
			self._pk = ed25519.Ed25519PublicKey.from_public_bytes(public_key.bytes)

	@property
	def crypto_scheme(self):
		"""Gets the crypto scheme identifier."""
		return self._scheme

	def verify(self, message, signature):
		"""Verifies a message signature."""
		if self._scheme == CryptoScheme.ED25519:
			try:
				self._pk.verify(signature.bytes, message)
				return True
			except InvalidSignature:
				return False

		backend = _get_pqc_backend(self._scheme)
		return backend.verify(message, signature.bytes, self.public_key.bytes)

File: sdk/javascript/src/symbol/KeyPairV2.js
import { PublicKeyV2, SignatureV2, PrivateKeyV2 } from './CryptoTypesV2.js';
import { CryptoScheme, PUBLIC_KEY_SIZES, PRIVATE_KEY_SIZES, SIGNATURE_SIZES } from './CryptoScheme.js';
import ed25519 from '../impl/ed25519.js';
import { deepCompare } from '../utils/arrayHelpers.js';

const HASH_MODE = 'Sha2_512';

// ML-DSA-65 key/signature constants
const ML_DSA_65_OQS_SECRET_KEY_SIZE = 4032;
const ML_DSA_65_PUBLIC_KEY_SIZE = 1952;

/**
 * Registry of PQC signing implementations.
 * The ML-DSA-65 backend must be registered externally (native addon or WASM).
 *
 * Expected backend interface:
 *   {
 *     generateKeyPair(): { publicKey: Uint8Array(1952), secretKey: Uint8Array(4032) }
 *     sign(message: Uint8Array, secretKey: Uint8Array): Uint8Array(3309)
 *     verify(message: Uint8Array, signature: Uint8Array, publicKey: Uint8Array): boolean
 *   }
 */
const pqcBackends = {};

/**
 * Registers a PQC signing backend for a given crypto scheme.
 * @param {number} schemeId CryptoScheme identifier.
 * @param {object} backend Implementation object with generateKeyPair/sign/verify.
 */
export const registerPqcBackend = (schemeId, backend) => {
	pqcBackends[schemeId] = backend;
};

/**
 * Gets the registered PQC backend for a given crypto scheme.
 * @param {number} schemeId CryptoScheme identifier.
 * @returns {object} Backend implementation.
 */
const getPqcBackend = schemeId => {
	const backend = pqcBackends[schemeId];
	if (!backend) throw new Error(`no PQC backend registered for scheme ${schemeId}; call registerPqcBackend() first`);
	return backend;
};

/**
 * Represents a V2 key pair supporting both Ed25519 and post-quantum schemes.
 */
export class KeyPairV2 {
	/**
	 * Creates a V2 key pair from a private key.
	 * @param {PrivateKeyV2} privateKey V2 private key (size determines scheme).
	 */
	constructor(privateKey) {
		if (!(privateKey instanceof PrivateKeyV2))
			throw new TypeError('privateKey must be a PrivateKeyV2 instance');

		this._scheme = privateKey.cryptoScheme;
		this._privateKey = privateKey;

		if (CryptoScheme.ED25519 === this._scheme) {
			this._keyPair = ed25519.get().keyPairFromSeed(HASH_MODE, this._privateKey.bytes);
		}
		// For PQC schemes, the public key is extracted from the stored private key format
	}

	/** Gets the crypto scheme identifier. */
	get cryptoScheme() {
		return this._scheme;
	}

	/**
	 * Gets the public key.
	 * @returns {PublicKeyV2} Public key.
	 */
	get publicKey() {
		if (CryptoScheme.ED25519 === this._scheme) {
			return new PublicKeyV2(this._keyPair.publicKey);
		}

		// ML-DSA-65: public key is stored at offset OQS_SECRET_KEY_SIZE in the private key
		const pkOffset = ML_DSA_65_OQS_SECRET_KEY_SIZE;
		return new PublicKeyV2(this._privateKey.bytes.subarray(pkOffset, pkOffset + ML_DSA_65_PUBLIC_KEY_SIZE));
	}

	/**
	 * Gets the private key.
	 * @returns {PrivateKeyV2} Private key.
	 */
	get privateKey() {
		return new PrivateKeyV2(this._privateKey.bytes);
	}

	/**
	 * Signs a message.
	 * @param {Uint8Array} message Message to sign.
	 * @returns {SignatureV2} Signature.
	 */
	sign(message) {
		if (CryptoScheme.ED25519 === this._scheme) {
			return new SignatureV2(ed25519.get().sign(HASH_MODE, message, this._keyPair));
		}

		const backend = getPqcBackend(this._scheme);
		const oqsSk = this._privateKey.bytes.subarray(0, ML_DSA_65_OQS_SECRET_KEY_SIZE);
		return new SignatureV2(backend.sign(message, oqsSk));
	}

	/**
	 * Generates a new V2 key pair for the given crypto scheme.
	 * @param {number} schemeId CryptoScheme identifier.
	 * @returns {Promise<KeyPairV2>} Generated key pair.
	 */
	static async generate(schemeId = CryptoScheme.ED25519) {
		if (CryptoScheme.ED25519 === schemeId) {
			const { randomBytes } = await import('crypto');
			const seed = randomBytes(32);
			return new KeyPairV2(new PrivateKeyV2(seed));
		}

		const backend = getPqcBackend(schemeId);
		const { publicKey: pk, secretKey: sk } = backend.generateKeyPair();

		// Store in our format: sk(4032) || pk(1952)
		const combined = new Uint8Array(PRIVATE_KEY_SIZES[schemeId]);
		combined.set(sk, 0);
		combined.set(pk, sk.length);
		return new KeyPairV2(new PrivateKeyV2(combined));
	}
}

/**
 * Verifies V2 signatures from any supported crypto scheme.
 */
export class VerifierV2 {
	/**
	 * Creates a V2 verifier from a public key.
	 * @param {PublicKeyV2} publicKey V2 public key.
	 */
	constructor(publicKey) {
		if (!(publicKey instanceof PublicKeyV2))
			throw new TypeError('publicKey must be a PublicKeyV2 instance');

		if (0 === deepCompare(new Uint8Array(publicKey.bytes.length), publicKey.bytes))
			throw new Error('public key cannot be zero');

		this.publicKey = publicKey;
		this._scheme = publicKey.cryptoScheme;
	}

	/** Gets the crypto scheme identifier. */
	get cryptoScheme() {
		return this._scheme;
	}

	/**
	 * Verifies a message signature.
	 * @param {Uint8Array} message Message to verify.
	 * @param {SignatureV2} signature Signature to verify.
	 * @returns {boolean} true if signature is valid.
	 */
	verify(message, signature) {
		if (CryptoScheme.ED25519 === this._scheme) {
			return ed25519.get().verify(HASH_MODE, message, signature.bytes, this.publicKey.bytes);
		}

		const backend = getPqcBackend(this._scheme);
		return backend.verify(message, signature.bytes, this.publicKey.bytes);
	}
}

export default {
	KeyPairV2,
	VerifierV2,
	registerPqcBackend
};
